src/shelly_light_bulb_controller.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) Shelly-HomeKit Contributors
+ * All rights reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mgos.hpp"
+
+#pragma once
+
+namespace shelly {
+
+class LightBulbControllerBase {
+ public:
+  enum class BulbType {
+    kBrightness = 0,
+    kColortemperature = 1,
+    kHueSat = 2,
+  };
+  typedef std::function<void()> Update;
+
+  LightBulbControllerBase(struct mgos_config_lb *cfg, Update ud);
+  LightBulbControllerBase(const LightBulbControllerBase &other) = delete;
+  virtual ~LightBulbControllerBase();
+
+  void UpdateOutput();
+
+  virtual BulbType Type() {
+    return BulbType::kBrightness;
+  }
+
+  bool IsOn() const;
+  bool IsOff() const;
+
+ protected:
+  struct mgos_config_lb *cfg_;
+
+  Update update_;
+};
+
+template <class T>
+class LightBulbController : public LightBulbControllerBase {
+ public:
+  LightBulbController(struct mgos_config_lb *cfg)
+      : LightBulbControllerBase(
+            cfg,
+            std::bind(&LightBulbController<T>::UpdateOutputSpecialized, this)),
+        transition_timer_(
+            std::bind(&LightBulbController<T>::TransitionTimerCB, this)) {
+  }
+  LightBulbController(const LightBulbControllerBase &other) = delete;
+
+ protected:
+  mgos::Timer transition_timer_;
+  int64_t transition_start_ = 0;
+
+  T state_start_;
+  T state_now_;
+  T state_end_;
+
+  virtual T ConfigToState() = 0;
+  virtual void ReportTransition(const T &next, const T &prev) = 0;
+  virtual void UpdatePWM(const T &state) = 0;
+
+  void TransitionTimerCB() {
```

Comment:
no code in headers please, unless it's trivial 1 or 2 liner.

Reply:
I have rewritten the code to use templates this weekend; this is to deduplicate the transition logic.

https://stackoverflow.com/questions/8752837/undefined-reference-to-template-class-constructor
there are only two ways to do this: either code in header or a priori instantiation;
see SO for details; would you prefer the other solution?